table/server/common/src/main/java/alluxio/table/common/udb/UdbAttachSpec.java

```diff
@@ -0,0 +1,612 @@
+
+  /**
+   * Tables and partitions to bypass.
+   */
+  private final TablePartitionInclusionExclusionWrapper mBypassed;
+
+  /**
+   * Ignored tables.
+   * Tables are ignored as a whole, no partition configuration is needed.
+   */
+  private final NamePatternInclusionExclusionWrapper mIgnored;
+
+  protected UdbAttachSpec(TablePartitionInclusionExclusionWrapper bypassed,
+                          NamePatternInclusionExclusionWrapper ignored) {
+    mBypassed = bypassed;
+    mIgnored = ignored;
+  }
+
+  /**
+   * Checks if a table should be bypassed.
+   *
+   * @param tableName the table name
+   * @return true if the table is configured to be bypassed, false otherwise
+   * @see UdbAttachSpec#hasFullyBypassedTable(String)
+   */
+  public boolean hasBypassedTable(String tableName) {
+    return shadowedByIgnore(tableName, mBypassed.hasTable(tableName));
+  }
+
+  /**
+   * Checks if all partitions of a table should be bypassed.
+   *
+   * @param tableName the table name
+   * @return true if the table is configured to be fully bypassed, false otherwise
+   * @see UdbAttachSpec#hasBypassedTable(String)
+   */
+  public boolean hasFullyBypassedTable(String tableName) {
+    return shadowedByIgnore(tableName, mBypassed.hasFullTable(tableName));
+  }
+
+  /**
+   * Checks if a table is ignored.
+   *
+   * @param tableName the table name
+   * @return true if the table is ignored, false otherwise
+   */
+  public boolean hasIgnoredTable(String tableName) {
+    return mIgnored.has(tableName);
+  }
+
+  /**
+   * Checks by a partition's name if it should be bypassed.
+   *
+   * @param tableName the table name
+   * @param partitionName the partition name
+   * @return true if the partition should be bypassed, false otherwise
+   */
+  public boolean hasBypassedPartition(String tableName, String partitionName) {
+    return shadowedByIgnore(tableName, mBypassed.hasPartition(tableName, partitionName));
+  }
+
+  /**
+   * Checks if a table is shadowed by ignoring when it is also bypassed.
+   * @param tableName the table name
+   * @param mightBeShadowedIfTrue the return value of {@link #hasBypassedTable(String)}, etc
+   * @return false if the table is shadowed by ignoring, mightBeShadowedIfTrue if it is not
+   */
+  private boolean shadowedByIgnore(String tableName, boolean mightBeShadowedIfTrue) {
+    if (mightBeShadowedIfTrue && hasIgnoredTable(tableName)) {
+      LOG.warn("Table {} is set to be bypassed but it is also ignored", tableName);
+      return false;
+    }
+    return mightBeShadowedIfTrue;
+  }
+
+  /**
+   * Wrapper class that takes into account inclusion and exclusion behaviour.
+   */
+  abstract static class InclusionExclusionWrapper<T extends NamePatternWrapper> {
+    protected final T mIncluded;
+    // excluded entries don't allow partition specifications -- they are name and regex only
+    protected final SimpleNamePatternWrapper mExcluded;
```

Comment:
> can you remind me why we don't allow the partition specifications in exclusion entry?

The design is to make the `partitions` entry itself a `InclusionExclusionWrapper`, so excluded partitions are specified with itself. If we allowed partition specification on the table level, e.g.

```json
{
    "bypass": {
        "tables": {
            "exclude": [
                {"table":"table1", "partitions": ["part1"]}
            ]
        }
    }
}
```

This can be confusing because one may wonder what exactly is excluded, `table1` or `part1`?

As opposed to the above, by prohibiting partition specification in exclusion list, we can be more clear on this:

```json
{
    "bypass": {
        "tables": {
            "include": [
                {
                    "table":"table1", 
                    "partitions": {
                        "exclude": ["part1"]
                    }
                }
            ]
        }
    }
}
```

This is the correct way to say "`table1` is included in bypassing, except for its partition `part1`".